### PR TITLE
Remove constants

### LINKED ISSUE #42: Harmonize `div` and `mul` between constants
`div` and `mul` between constants are optimized by the compiler and does not cost more gas than using constants. Perhaps we should come back to `div(MAX_UINT256, HALF_PERCENTAGE_FACTOR)`?

### PR DESCRIPTION
**Pull request by MerlinEgalite**
Fixes #42

### CODE REVIEW

File: src/math/PercentageMath.sol [hunk 1]
```diff
@@ -10,8 +10,6 @@ library PercentageMath {
 
     uint256 internal constant PERCENTAGE_FACTOR = 1e4; // 100.00%
```
Comment (MathisGD) [inline src/math/PercentageMath.sol:11]: IMO, this name is not perfect.. But I think that we want to have the same interface than Aave here right ?

File: src/math/PercentageMath.sol [hunk 2]
```diff
@@ -10,8 +10,6 @@ library PercentageMath {
...
     uint256 internal constant HALF_PERCENTAGE_FACTOR = 0.5e4; // 50.00%
```
Comment (MathisGD) [inline src/math/PercentageMath.sol:12]: What is the reason behind not using `div(PERCENTAGE_FACTOR, 2)`

File: src/math/WadRayMath.sol [hunk 3]
```diff
@@ -14,9 +14,6 @@ library WadRayMath {
     uint256 internal constant HALF_RAY = 0.5e27;
```
Comment (MathisGD) [inline src/math/WadRayMath.sol:14]: same question here

File: src/math/PercentageMath.sol (hunk 1 above)
Comment (MerlinEgalite) [reply src/math/PercentageMath.sol:11]: yes

Comment (MathisGD) [reply src/math/PercentageMath.sol:11]: Ok

File: src/math/PercentageMath.sol (hunk 2 above)
Comment (MerlinEgalite) [reply src/math/PercentageMath.sol:12]: because Aave is already doing it, so I  guess we can keep it

File: src/math/WadRayMath.sol (hunk 3 above)
Comment (MerlinEgalite) [reply src/math/WadRayMath.sol:14]: same

File: src/math/CompoundMath.sol [hunk 4]
```diff
@@ -18,7 +17,7 @@ library CompoundMath {
         assembly {
             // Revert if x * y > type(uint256).max
             // <=> y > 0 and x > type(uint256).max / y
-            if mul(y, gt(x, div(MAX_UINT256, y))) {
```
[the remark this answers is not in the document]
Comment (MerlinEgalite) [reply src/math/CompoundMath.sol:21]: yup we can change it

File: src/math/PercentageMath.sol (hunk 2 above)
Comment (MerlinEgalite) [reply src/math/PercentageMath.sol:12]: I mean we should at least expose the same function and getters

Comment (MathisGD) [reply src/math/PercentageMath.sol:12]: ok ok

**Review by Rubilmax — COMMENTED** [review]
The thing is, these libraries could be used with older versions of Solidity, if ever needed
And nothing ensures previous versions of the Solidity compiler did such constants optimizations (`sub(constant, constant) => constant`)

Also, did we check if one benefits from such optimizations without the optimization setting of the compiler?

TLDR; we should lower the solidity pragma of these libraries and don't rely on the compiler to optimize our code (just for the sake of beautiful code with less explicit constants) => we should explicitely define all such "optimization" implicit constants

**Comment by MerlinEgalite** [comment]
> The thing is, these libraries could be used with older versions of Solidity, if ever needed And nothing ensures previous versions of the Solidity compiler did such constants optimizations (`sub(constant, constant) => constant`)
> 
> Also, did we check if one benefits from such optimizations without the optimization setting of the compiler?
> 
> TLDR; we could lower the solidity pragma of these libraries and we shouldn't rely on the compiler to optimize our code (just for the sake of beautiful code with less explicit constants) => we should explicitely define all such "optimization" implicit constants

This is a good point. We might not merge this PR then.

**Comment by MathisGD** [comment]
I don't think that relying on newest versions of solidity for some optimizations is not an issue. It is quite common to simplify you code when the solidity compiler changes its behavior (see Solmate, relying on solidity integrated overflow management). Moreover, the number of people developing with really old versions of solidity is very small.

**Comment by MerlinEgalite** [comment]
Yes I didn't the see the bots. Seems not to be an improvement, we must check in local too.

Good points too Mathis. Unless this discussion is resolved, I'll keep this PR pending.

**Comment by MathisGD** [comment]
Ok so the gas cost is worse than before if the number of optimizations of the compiler is low...

**Comment by MathisGD** [comment]
After thinking about it and discussing with part of the team, I think that we should optimize our code for a high number of optimization. So I would change the default number of optimizations of this repo, and merge it.

**Review by Rubilmax — DISMISSED** [review]
What's the point in making these changes?

We're making optimization implicit, for the sake of the beauty of the code. I disagree with this move

**Comment by MathisGD** [comment]
Readability (which helps maintaining a safe code)

**Comment by MathisGD** [comment]
What should we do with this one ?

File: src/math/CompoundMath.sol [hunk 5]
```diff
@@ -8,7 +8,7 @@ pragma solidity ^0.8.0;
 library CompoundMath {
     /// CONSTANTS ///
 
-    uint256 internal constant MAX_UINT256 = 2**256 - 1;
+    uint256 internal constant MAX = 2**256 - 1;
     uint256 public constant SCALE = 1e36;
```
Comment (pakim249CAL) [inline src/math/CompoundMath.sol:12]: Unrelated, but maybe we should be declaring these as internal too. Declaring a library variable as public would technically make this an external library.

**Review by pakim249CAL — APPROVED** [review]

**Comment by Rubilmax** [comment]
I still disagree with this change, as I don't think removing explicit constants increases the readability of the code (it even decreases it IMO) lol

**Comment by MathisGD** [comment]
Are you sure that you prefer
```solidity
MAX_MINUS_HALF_RAY
```

over

```solidity
sub(MAX, div(RAY, 2))
```

It is not obvious to me
